File: 24. Swap Nodes in Pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (head == NULL or head->next == NULL) return head;
        ListNode* dummy = new ListNode(-1);
        ListNode* p = dummy, *p1 = head, *p2 = p1->next;
        while (p1 != NULL && p2 != NULL)
        {
            p->next = p2;
            p1->next = p2->next;
            p2->next = p1;
            p = p1;
            p1 = p1->next;
            if (p1 != NULL) p2 = p1->next;
        }
        return dummy->next;
    }
};

File: 24. Swap Nodes in Pairs.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return head
        dummy = ListNode(-1)
        p = dummy
        p1 = head
        p2 = p1.next
        while p1 and p2:
            p.next = p2
            p1.next = p2.next
            p2.next = p1
            p = p1
            p1 = p1.next
            if p1:
                p2 = p1.next
        return dummy.next
